src/lib/provable/crypto/foreign-twisted.ts

```diff
@@ -0,0 +1,309 @@
+import {
+  TwistedCurveParams,
+  createCurveTwisted,
+} from '../../../bindings/crypto/elliptic-curve.js';
+import type { Group } from '../group.js';
+import { ProvablePureExtended } from '../types/struct.js';
+import { AlmostForeignField, createForeignField } from '../foreign-field.js';
+import { CurveTwisted, Point } from '../gadgets/twisted-curve.js';
+import { Field3 } from '../gadgets/foreign-field.js';
+import { assert } from '../gadgets/common.js';
+import { Provable } from '../provable.js';
+import { provableFromClass } from '../types/provable-derivers.js';
+import { l2Mask, multiRangeCheck } from '../gadgets/range-check.js';
+import { Bytes } from '../bytes.js';
+
+// external API
+export { createForeignTwisted, ForeignTwisted };
+
+// internal API
+export { toPoint, FlexiblePoint };
+
+type FlexiblePoint = {
+  x: AlmostForeignField | Field3 | bigint | number;
+  y: AlmostForeignField | Field3 | bigint | number;
+};
+function toPoint({ x, y }: ForeignTwisted): Point {
+  return { x: x.value, y: y.value };
+}
+
+class ForeignTwisted {
+  x: AlmostForeignField;
+  y: AlmostForeignField;
+
+  /**
+   * Create a new {@link ForeignTwisted} from an object representing the (twisted) x and y coordinates.
+   *
+   * Note: Inputs must be range checked if they originate from a different field with a different modulus or if they are not constants. Please refer to the {@link ForeignField} constructor comments for more details.
```

Comment:
Nit: but it is in the entire o1js codebase - could we wrap at 80 characters please? It is painful when you want to split your window in different buffers.

Reply:
I would agree to set a 100-char limit. However, my IDE automatically wraps adjusting to the tab width, so I never find this annoying personally.